flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricQueryService.java

```diff
@@ -119,6 +121,101 @@ public void onReceive(Object message) {
 		}
 	}
 
+	public void unregister(Metric metric) {
+		if (metric instanceof Counter) {
+			this.counters.remove(metric);
+		} else if (metric instanceof Gauge) {
+			this.gauges.remove(metric);
+		} else if (metric instanceof Histogram) {
+			this.histograms.remove(metric);
+		} else if (metric instanceof Meter) {
+			this.meters.remove(metric);
+		}
+	}
+
+	private MetricDumpSerialization.MetricSerializationResult verifyResultAndUnregisterOversizedMetrics(
+		MetricDumpSerialization.MetricSerializationResult serializationResult) {
+
+		byte[] serializedCounters;
+		int numCounters;
+		if (serializationResult.serializedCounters.length > maximumFramesize) {
+			LOG.warn("The serialized counter metric is larger than the maximum frame size, " +
+				" so maybe not all metrics would be reported.");
+			serializedCounters = new byte[0];
+			numCounters = 0;
+
+			for (Metric metric : this.counters.keySet()) {
+				unregister(metric);
```

Comment:
So i was thinking about this for a while now. Un-registering metrics because they don't fit may be a bad idea after all.

These metrics will never be registered again; they are lost, potentially due to a some short-term issue with the # of metrics. In other words, _some_ short-term issue affects the cluster in a permanent way and can only be resolved by restarting it.

Let's not un-register them for now. We pay the _kind of unnecessary_ serialization cost, but that's fine for now. The PR improves the current situation anyway; we can handle this kind of optimizations in a follow-up.

WDYT?

Reply:
@zentol I agree with you. Oversize may be just a short-term phenomenon (in particular, `maximum-frame-size` is still configurable), the unregister operation may be overreacting. Subsequent optimizations can use statistical oversize times to estimate whether this phenomenon is long-term.